### PR TITLE
Implement request to undo moves for #20

### LINKED ISSUE #20: Take back/undo moves
Allow players to request a take-back. Will require pulling popping old game states off the stack, and a new button + pop-up.

Frontend and backend work.

### PR DESCRIPTION
**Pull request by rovmug-ticfyn**
This partially resolves #20, only backend at this point

### CODE REVIEW

File: src/urbit/sur/historic.hoon
Comment (ashelkovnykov) [inline src/urbit/sur/historic.hoon]: Should be `chess-game-0`, to denote when it was in the active state

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: `active-game-state-0` should be in `historic.hoon` as well

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: This should be `chess-game-0.historic` (or possibly `chess-game-0`: see below comment)

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: Can add `=,  historic` below this line so that types don't need to be suffixed by `.historic`. This should work in all cases, as the historic types end with `-#` where `#` denotes in what game state they were active.

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: This is a solid initial draft of parsing a FEN to a `chess-position`. However:
1. Pinning a named value to the subject and editing it is a natural response a lot of programmers have when coming to Hoon, but it's something that should be avoided as often as possible. The reason is that since Hoon is a functional language, it can't actually append or modify the subject/state: it can only clone existing state with modifications. This means that each state update is creating a copy of some subtree of the subject, which is inefficient, and thus should be avoided as much as possible.
2. Are you familiar with the `&n` and `|n` operators? `&n` produces the `nth` value of a tuple with at least `n + 1` entries. `|n` produces the value directly after `&n`. Example:
    ```
    > =y ['a' 'b' 'c']
    > &2.y
    'b'
    > |2.y
    'c'
    ```
3. It's a little odd that some things are parsed during splitting and others are parsed after splitting.
4. `parsed-fen` is somewhat of a misnomer, given that additional parsing happens on its data.

In addition to the above, there's some efficiency and correctness concerns that it would take a long time to patch going back and forth together. I recommend replacing the above implementation with this one:
```
++  fen-to-position
|=  fen=@t
^-  chess-position
%+  rush
  fen
|.
;~  (glue ace)
    fen-to-board         ::  board
    fen-to-player        ::  player-to-move
    fen-to-white-castle  ::  white-can-castle
    fen-to-black-castle  ::  black-can-castle
    fen-to-en-passant    ::  en-passant
    dem                  ::  ply-50-move-rule
    dem                  ::  move-number
==
++  fen-to-board
  %+  cook  board-helper
  %+  more  (just '/')
  ::  XX is this actually slower than alf and fail on invalid char?
  (stun [1 8] (mask "12345678bknpqrBKNPQR"))
++  fen-to-player
  ;~  pose
      (cold %white (just 'w'))
      (cold %black (just 'b'))
  ==
++  fen-to-white-castle
  ;~  pose
      (white-castle-helper hep " -" %none)
      (white-castle-helper (jest 'k') " k" %none)
      (white-castle-helper (jest 'q') " q" %none)
      (white-castle-helper (jest 'KQ') " " %both)
      (white-castle-helper (just 'K') " " %kingside)
      (white-castle-helper (just 'Q') " " %queenside)
  ==
++  fen-to-black-castle
  ;~  pose
      (cold %none hep)
      (cold %both (jest 'kq'))
      (cold %kingside (just 'k'))
      (cold %queenside (just 'q'))
  ==
++  fen-to-en-passant
  ;~  pose
      (cold ~ hep)
      %+  cook
          some
          ;~  plug
              (cook term (shim 'a' 'h'))
              (cook term (mask "36"))
          ==
  ==
++  board-helper
  |=  brd=(list tape)
  ?>  =((lent brd) 8)
  =/  rank  %8
  =/  file  %a
  =/  board  *(map chess-square chess-piece)
  |-
  ?~  brd
    board  
  ?<  (gth file %g)
  ?:  =(file %g)
    $(brd t.brd, rank (sub rank 1), file %a)
  ?+  (term  (cut 1 [(sub file 97) 3] i.brd))  !!
    %1  $(file (add file 1))
    %2  $(file (add file 2))
    %3  $(file (add file 3))
    %4  $(file (add file 4))
    %5  $(file (add file 5))
    %6  $(file (add file 6))
    %7  $(file (add file 7))
    %8  $(file (add file 8))
    ::
    %b  $(file +(file), board (~(put by board) [[file rank] [%black %bishop]]))
    %k  $(file +(file), board (~(put by board) [[file rank] [%black %king]]))
    %n  $(file +(file), board (~(put by board) [[file rank] [%black %knight]]))
    %p  $(file +(file), board (~(put by board) [[file rank] [%black %pawn]]))
    %q  $(file +(file), board (~(put by board) [[file rank] [%black %queen]]))
    %r  $(file +(file), board (~(put by board) [[file rank] [%black %rook]]))
    ::
    %B  $(file +(file), board (~(put by board) [[file rank] [%white %bishop]]))
    %K  $(file +(file), board (~(put by board) [[file rank] [%white %king]]))
    %N  $(file +(file), board (~(put by board) [[file rank] [%white %knight]]))
    %P  $(file +(file), board (~(put by board) [[file rank] [%white %pawn]]))
    %Q  $(file +(file), board (~(put by board) [[file rank] [%white %queen]]))
    %R  $(file +(file), board (~(put by board) [[file rank] [%white %rook]]))
  ==
++  white-castle-helper
  |=  [rul=rule pre=tape res=@tas]
  ^-  rule
  ;~(pfix rul (funk pre (easy res)))
```
Some notes about this implementation:
1. I didn't test it out, on anything: it's written entirely by intuition, just so that I don't take even longer on this PR and can get to the other ones. I'm hoping it's close enough to correct that I'm not laying a frustrating task onto you to patch it into its final state.
2. Despite catching more "bad FEN" cases, this is still an "unsafe" conversion function, since it doesn't check for a "well-formed" FEN. What we want is two `fen-to-position` functions: one safe, and one unsafe. The safe one can be used when we are converting FENs that we ourselves created back to `chess-position`s. The unsafe one is for importing FENs from externally. The unsafe version of the function should call the safe version, and then perform additional validation. Safe `fen-to-position` validation checks would include:
  a. exactly one king per side
  b. no more than 16 pieces per side
  c. no more than 10 rooks, bishops, or knights per side
  d. no more than 9 queens per side
  e. no more than 8 pawns per side
  f. no pawns on back ranks
  g. en passant square makes sense for board arrangement
  h. castle values make sense for board arrangement

Note that the safe version of `fen-to-position` is not necessary at this time. We can just add an empty arm with a comment about it, like `fen-to-position` had before this PR.

File: src/urbit/mar/chess/undo-accept.hoon
Comment (ashelkovnykov) [inline src/urbit/mar/chess/undo-accept.hoon]: Add a new line at the end of each mark file

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon:444]: This check should be relaxed: most often, the reason that someone requests an undo is because he makes a move that becomes apparent as a blunder after the opponent responds.

- If we're white, it's turn 25, it's our move, and we request an undo: we're trying to go back to our own move on turn 24.
- If we're white, it's turn 25, it's black's move, and we request an undo: we're tying to go back to our own move on turn 25.
- If we're black, it's turn 25, it's white's move, and we request an undo: we're trying to go back to our own move on turn 24.
- If we're black, it's turn 25, it's our move, and we request an undo: we're trying to go back to our own move on turn 24.

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon:452]: Need to send out an update on the `/game/../updates` wire so that the frontend can visualize the undo process

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon:479]: Same here, re: `update` wire card

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: Ugh - having to pin intermediate values to the subject just to update the state is inefficient, but there's no better way to do it here. I tried to write an alternative which updates the entire state in one go, and it's so hideously convoluted that it's probably better for the sake of readability to keep it like you wrote. Only recommendation I have is to merge the state updates using `=:` and to implement the undo functionality I described in one of the above comments:
```
=/  ship-to-move
  ?-  player-to-move.position.u.game-state
    %white  white.game.u.game-state
    %black  black.game.u.game-state
  ==
?>  ?=([%ship @p] ship-to-move)
...
=:
  moves.game.u.game  ?:(=(+.ship-to-move our) (snip moves.game.u.game) (snip (snip moves.game.u.game)))
  position.u.game (fen-to-position +.(rear moves.game.u.game))
  got-undo-request.u.game |
==
```

Unrelated:
1. These `ship-to-move` calls are really inefficient and unnecessary. This should just be part of the game state.
2. The way we handle moves is incredibly inefficient. We should be head-appending moves, instead of using `snoc`/`snip`. The frontend can reverse the list itself for in-progress games, otherwise it can build record moves in an array however it needs from moves along the `update` wire in a single session.

Those two points above are not actionable items for you, I'm just noting this down for posterity.

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: This isn't a bad idea, actually, if it were set to `=,  u.game-state`. This could be useful in the `%accept-undo` block above too: rename `game` to `game-state` and add this block after checking that `game-state` is non-null.

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: Same advice as above, but invert the relationship based on whose turn it is. Combined with the preceding comment about using `=,`:
```
moves.game  ?:(=(+.ship-to-move our) (snip (snip moves.game)) (snip moves.game))
```

**Review by ashelkovnykov — CHANGES_REQUESTED** [review]
What you have here is entirely workable - there are just some optimizations available to use that we should use.

File: src/urbit/lib/chess.hoon
Comment (rovmug-ticfyn) [reply src/urbit/lib/chess.hoon]: Based on [this post](https://developers.urbit.org/blog/text-overview), a term can't contain a capital letter, so we'll need a different way to differentiate the black and white pieces in `board-helper`. 

Maybe we first perform a `(sane @tas)` check. If yes, handle it with the `?+` you've written. If no, a series of `?:`s that check for the white pieces?

Comment (ashelkovnykov) [reply src/urbit/lib/chess.hoon]: Ooph - bummer, dude. Let me think on this!

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Alright, here's a wild one for ya - replace these lines with:
```
?:  (lte next 56)
  $(file (chess-file (add file (sub next 48))))
?:  (gte next 97)
  ?+  (term next)  !!
    %b  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %bishop]]))
    %k  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %king]]))
    %n  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %knight]]))
    %p  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %pawn]]))
    %q  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %queen]]))
    %r  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%black %rook]]))
  ==
?+  (term (add next 32))  !!
  %b  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %bishop]]))
  %k  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %king]]))
  %n  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %knight]]))
  %p  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %pawn]]))
  %q  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %queen]]))
  %r  $(file (chess-file +(file)), board (~(put by board) [[file rank] [%white %rook]]))
==
```
This is trying to use more [ASCII magic](https://www.asciitable.com) to convert a text character into a literal which Hoon can interpret as a type for `?+`. We also need to explicitly mold `file` with `chess-file` so that `next-file` isn't necessary (Look out for more comments about this).

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: All of these lines should be indented once more

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Okay, I made some mistakes in my first draft (including forgetting that `%g` comes *before* `%h`, not after it...). Try this:
```
=/  rank=chess-rank  %8
=/  file=$?(chess-file %i)  %a
...
  board
?-  file
    %i
  $(brd t.brd, rank (chess-rank (sub rank 1)), file %a)
::
    chess-file
  =*  next  (snag (sub file 97) i.brd)
  ...
==
```

**Review by ashelkovnykov — COMMENTED** [review]
Okay, I think I've got some fixes for our parsing problem

**Comment by ashelkovnykov** [comment]
@rovmug-ticfyn Whatever you did when adding your latest batch of changes has really confused GitHub. Could you crush all of your changes into one commit that's rebased directly onto the master branch? Hopefully that will fix it - otherwise, it's just impossible to review right now because GitHub is showing your changes intermixed with a ton of changes we've already merged.

File: src/urbit/sur/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/sur/chess.hoon]: We should probably declare a `chess-fen` type which just maps to `@t`, so that the type is self-documenting

File: src/urbit/sur/historic.hoon
Comment (ashelkovnykov) [inline src/urbit/sur/historic.hoon]: Need newline at end of file

File: src/urbit/sur/historic.hoon
Comment (ashelkovnykov) [inline src/urbit/sur/historic.hoon:2]: Missing `state-0`

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: These two lines should be indented two more

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: As commented above, should be in `historic.hoon`

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: Might be a bug here: just like we have to snip either two moves or one based on whose turn it is for `moves.game`, we need to get the FEN of either the most recent move or the second-most recent move

File: src/urbit/app/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/app/chess.hoon]: Same as above

**Review by ashelkovnykov — CHANGES_REQUESTED** [review]
Let me know if you want me to take a look at what's causing the `nest-fail` error

**Comment by rovmug-ticfyn** [comment]
Getting a `nest-fail` error from the `fen-to-position` parsing telling me we have a nail that shouldn't be there (`[p=[p=@ud q=@ud] q=""]`). I'm still working on debugging, but no progress yet.

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: These two should be `just`, not `jest`

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Should be `res=chess-castle`

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: This line should be indented two spaces fewer, but with two more spaces before file:
```
  ?-    file
```

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: These lines should be indented two spaces more

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Should be `=/  board  *chess-board`

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: This gate is the root cause of the `nest-fail` error - both uses of `rule` are too restrictive, so the compiler accidentally encodes very strict types for what a "rule" is at the point in time when we call it. Code should instead be:
```
|*  [rul=rule pre=tape res=chess-castle]
;~(pfix rul (funk pre (easy res)))
```

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Need explicit casts to `chess-file` and `chess-rank`:
```
(cook chess-file (cook term (shim 'a' 'h')))
(cook chess-rank (cook term (mask "36")))
```

File: src/urbit/lib/chess.hoon
Comment (ashelkovnykov) [inline src/urbit/lib/chess.hoon]: Not used

**Review by ashelkovnykov — CHANGES_REQUESTED** [review]
Added comments which fix the `nest-fail` issue. Also, branch needs to be rebased onto most recent master. While looking at the PR, I made a branch that did both of these things, if you'd like to avoid repeating the same work. It's available [here](https://github.com/ashelkovnykov/urbit-chess/tree/undo).

Note that the code still does not compile after this - the `algebraicize` functions throw errors about missing the FEN (which we've added to the list of moves).

**Review by ashelkovnykov — COMMENTED** [review]
Looks good to me. Stick a frontend on this and we're good!

**Review by ashelkovnykov — APPROVED** [review]
Approving PR to unblock other issues; frontend can come in a separate issue. Please squash commits before merge.

**Review by ashelkovnykov — APPROVED** [review]
LGTM

**Pull request merged**
